tests/DecisionServiceTests/DecisionServiceTest.php

```diff
@@ -603,6 +614,445 @@ public function testGetVariationWithBucketingId()
+
+        $this->decisionServiceMock = $this->getMockBuilder(DecisionService::class)
+        ->setConstructorArgs(array($this->loggerMock, $this->config))
+        ->setMethods(array('getVariationForFeatureExperiment','getVariationForFeatureRollout'))
+        ->getMock();
+
+        $feature_flag = $this->config->getFeatureFlagFromKey('string_single_variable_feature');
+        $rollout_id = $feature_flag->getRolloutId();
+        $rollout = $this->config->getRolloutFromId($rollout_id);
+        $experiment = $rollout->getExperiments()[0];
+        $expected_variation = $experiment->getVariations()[0];
+        $expected_decision = [
+            'experiment' => null,
+            'variation' => $expected_variation
+        ];
+
+        $this->decisionServiceMock
+        ->method('getVariationForFeatureExperiment')
+        ->will($this->returnValue(null));
+
+        $this->decisionServiceMock
+        ->method('getVariationForFeatureRollout')
+        ->will($this->returnValue($expected_variation));
+
+        $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::INFO, 
+            "User 'user_1' was bucketed into a rollout for feature flag 'string_single_variable_feature'.");
+
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeature($feature_flag, 'user_1', []),
+            $expected_decision
+        );
+    }
+
+    public function testGetVariationForFeatureWhenTheUserIsNeitherBucketedIntoFeatureExperimentNorToFeatureRollout(){
+        $this->decisionServiceMock = $this->getMockBuilder(DecisionService::class)
+        ->setConstructorArgs(array($this->loggerMock, $this->config))
+        ->setMethods(array('getVariationForFeatureExperiment','getVariationForFeatureRollout'))
+        ->getMock();
+
+        $feature_flag = $this->config->getFeatureFlagFromKey('string_single_variable_feature');
+
+         $this->decisionServiceMock
+        ->method('getVariationForFeatureExperiment')
+        ->will($this->returnValue(null));
+
+        $this->decisionServiceMock
+        ->method('getVariationForFeatureRollout')
+        ->will($this->returnValue(null));
+        
+        $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::INFO, 
+            "User 'user_1' was not bucketed into a rollout for feature flag 'string_single_variable_feature'.");
+
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeature($feature_flag, 'user_1', []),
+            null
+        );
+    }
+
+    public function testGetVariationForFeatureRolloutWhenNoRolloutIsAssociatedToFeatureFlag(){
+         $feature_flag = $this->config->getFeatureFlagFromKey('boolean_feature');
+
+         $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::DEBUG, 
+            "Feature flag 'boolean_feature' is not used in a rollout.");
```

Comment:
in any rollout? what does make more sense 'a' or 'any'?

Reply:
jo bhi pehle likh dia jaye 🗡